File: apps/host/internal/automationhost/service.go
// Package automationhost assembles the Host's production scheduling path: an
// owned Rust command Worker, the frozen session definitions the Host rebuilds
// on it, the persistent automation engine and the payload/authorization
// lookups a dispatch is re-checked against. It opens no listener of its own and
// never writes into an interactive PTY.
package automationhost

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"path/filepath"
	"regexp"
	"sync"
	"time"

	pb "armadra.local/host/gen/armadra/v1"
	"armadra.local/host/internal/automation"
	"armadra.local/host/internal/commanddispatch"
	auth "armadra.local/host/internal/identity"
	"armadra.local/host/internal/storage"
	"armadra.local/host/internal/worker"
	"google.golang.org/protobuf/proto"
)

var (
	// ErrUnsupported means this Host was started without an execution Worker,
	// or the request asks for a capability it does not have. It is never
	// answered with empty data pretending the operation succeeded.
	ErrUnsupported = errors.New("automation execution is unavailable on this Host")
	ErrInvalid     = automation.ErrInvalid
	idPattern      = regexp.MustCompile(`^[A-Za-z0-9_.-]{1,128}$`)
)

const (
	// Scope names checked against the verified session's recorded grants.
	ScopeRead   = "automation:read"
	ScopeManage = "automation:manage"

	defaultMaxRestarts   = 5
	defaultRestartBase   = time.Second
	defaultRestartCap    = 30 * time.Second
	defaultHealthyAfter  = time.Minute
	defaultRequestBudget = 10 * time.Second
	rebuildPage          = 200
)

type Options struct {
	// Executable and StateDir are explicit configuration. An unset pair means
	// automation is not enabled; a half-configured pair is a startup error.
	Executable, StateDir string
	HostID, InstanceID   string
	Store                *storage.Store
	// Sessions is the session domain's landing point for a run report. Nil on a
	// Host that assembles no session service: the frame is still recorded, it
	// simply changes no session record.
	Sessions SessionObserver
	// Agents is the agent domain's landing point for a Hook report, on the same
	// terms.
	Agents AgentObserver
	Clock                func() time.Time
	RequestTimeout       time.Duration
	PollInterval         time.Duration
	MaxRestarts          int
	RestartBackoff       time.Duration
	HealthyAfter         time.Duration
}

// Caller is the verified session's own identity and grants. Nothing here comes
// from a client-supplied field.
type Caller struct {
	PrincipalID, DeviceID, WorkspaceID string
	DeviceEpoch                        uint64
	Scopes                             []auth.Scope
}

type Service struct {
	options   Options
	store     *storage.Store
	engine    *automation.Engine
	clock     func() time.Time
	mu        sync.Mutex
	client    *worker.Client
	inner     *commanddispatch.Dispatcher
	startedAt time.Time
	reason    string
	closing   sync.Once
	closeErr  error
	closed    bool
}

// Configured reports whether the operator asked for an execution Worker at all.
func Configured(options Options) bool { return options.Executable != "" || options.StateDir != "" }

// New starts the Worker, rebuilds the stored command definitions on it and
// prepares the engine. It does not begin scheduling; Run does.
func New(ctx context.Context, options Options) (*Service, error) {
	if options.Executable == "" || options.StateDir == "" || options.Store == nil {
		return nil, ErrUnsupported
	}
	if !filepath.IsAbs(options.Executable) || !filepath.IsAbs(options.StateDir) {
		return nil, errors.New("automation requires absolute Worker executable and state directory paths")
	}
	if options.Clock == nil {
		options.Clock = time.Now
	}
	if options.RequestTimeout == 0 {
		options.RequestTimeout = defaultRequestBudget
	}
	if options.MaxRestarts == 0 {
		options.MaxRestarts = defaultMaxRestarts
	}
	if options.RestartBackoff == 0 {
		options.RestartBackoff = defaultRestartBase
	}
	if options.HealthyAfter == 0 {
		options.HealthyAfter = defaultHealthyAfter
	}
	s := &Service{options: options, store: options.Store, clock: options.Clock, reason: "WORKER_NOT_STARTED"}
	engine, err := automation.New(options.Store, s, s, automation.Options{Clock: options.Clock, InstanceID: options.InstanceID, PollInterval: options.PollInterval})
	if err != nil {
		return nil, err
	}
	s.engine = engine
	// Runs written before the history index existed are projected into it once,
	// before anything can page them: a run history that silently began at the
	// upgrade would look truncated rather than ordered.
	if err = engine.EnsureRunHistory(ctx); err != nil {
		return nil, err
	}
	if err = s.start(ctx); err != nil {
		return nil, err
	}
	return s, nil
}

// Run owns the Host lifetime, never a page or request. It returns only after
// both the schedule loop and Worker supervision have stopped.
func (s *Service) Run(parent context.Context) error {
	ctx, cancel := context.WithCancel(parent)
	defer cancel()
	supervision := make(chan struct{})
	go func() { defer close(supervision); s.supervise(ctx) }()
	err := s.engine.Run(ctx)
	cancel()
	<-supervision
	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
		return nil
	}
	return err
}

// Close stops the owned Worker and reports whether its children were reclaimed.
// Callers stop Run first: a Worker must not be shut down under a live dispatch.
func (s *Service) Close() error {
	if s == nil {
		return nil
	}
	s.closing.Do(func() {
		s.mu.Lock()
		client := s.client
		s.client, s.inner, s.closed, s.reason = nil, nil, true, "HOST_STOPPING"
		s.mu.Unlock()
		s.closeErr = client.Close()
	})
	return s.closeErr
}

func (s *Service) now() int64 { return s.clock().UnixMilli() }

func (s *Service) current() (*worker.Client, *commanddispatch.Dispatcher, time.Time) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.client, s.inner, s.startedAt
}

// Unavailable reports the recorded reason no Worker is currently serving, or
// an empty string while one is live. It is diagnostic, not an authorization.
func (s *Service) Unavailable() string {
	if s == nil {
		return "AUTOMATION_NOT_CONFIGURED"
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.client != nil {
		return ""
	}
	return s.reason
}

func (s *Service) start(ctx context.Context) error {
	s.mu.Lock()
	closed := s.closed
	s.mu.Unlock()
	if closed {
		return ErrUnsupported
	}
	// Every rebuilt Worker gets the upcall sink, so the supervisor's existing
	// restart path is also the channel's reconnect path: a Worker that comes
	// back replays what it owes over the new pipe with no extra machinery.
	client, err := worker.Start(ctx, worker.Options{Executable: s.options.Executable, HostID: s.options.HostID, StateDir: s.options.StateDir, RequestTimeout: s.options.RequestTimeout, Upcalls: upcallRecorder{store: s.store, hostInstance: s.options.InstanceID, sessions: s.options.Sessions, agents: s.options.Agents}})
	if err != nil {
		s.disable("WORKER_START_FAILED")
		return err
	}
	inner, err := commanddispatch.New(client, payloads{s.store})
	if err != nil {
		s.disable("WORKER_COMMANDS_UNAVAILABLE")
		return errors.Join(err, client.Close())
	}
	if err = s.rebuild(ctx, client); err != nil {
		s.disable("DEFINITION_REBUILD_FAILED")
		return errors.Join(err, client.Close())
	}
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return errors.Join(ErrUnsupported, client.Close())
	}
	s.client, s.inner, s.startedAt, s.reason = client, inner, s.clock(), ""
	s.mu.Unlock()
	return nil
}

func (s *Service) disable(reason string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.closed {
		s.client, s.inner, s.reason = nil, nil, reason
	}
}

func backoff(attempt int, base, limit time.Duration) time.Duration {
	wait := base
	for range attempt {
		if wait >= limit/2 {
			return limit
		}
		wait *= 2
	}
	return wait
}

func wait(ctx context.Context, delay time.Duration) bool {
	timer := time.NewTimer(delay)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

// supervise restarts an unexpectedly exited Worker a bounded number of times.
// A restart re-applies the stored definitions; it never re-sends a dispatch,
// which stays keyed by its stable operation ID and its durable receipt.
func (s *Service) supervise(ctx context.Context) {
	attempts := 0
	for {
		client, _, startedAt := s.current()
		if client != nil {
			select {
			case <-ctx.Done():
				return
			case <-client.Done():
			}
			_ = client.Close()
			s.disable("WORKER_EXITED")
			if s.clock().Sub(startedAt) >= s.options.HealthyAfter {
				attempts = 0
			}
		}
		if ctx.Err() != nil {
			return
		}
		if attempts >= s.options.MaxRestarts {
			s.disable("WORKER_RESTART_BUDGET_EXHAUSTED")
			return
		}
		if !wait(ctx, backoff(attempts, s.options.RestartBackoff, defaultRestartCap)) {
			return
		}
		attempts++
		if err := s.start(ctx); err != nil {
			continue
		}
	}
}

// rebuild re-applies the Host's own stored definitions to a fresh Worker and
// verifies the generation it reports. A definition that cannot be rebuilt, or
// that comes back with a different generation or launch digest, is recorded as
// unrebuildable so plans targeting it are refused instead of waiting forever.
func (s *Service) rebuild(ctx context.Context, client *worker.Client) error {
	roots, err := s.store.CommandRoots(ctx)
	if err != nil {
		return err
	}
	bound := map[string]bool{}
	for _, root := range roots {
		if _, err = client.BindCommandRoot(ctx, &pb.BindCommandRootRequest{RootId: root.RootID, WorkspaceId: root.WorkspaceID, Path: root.Path}); err == nil {
			bound[root.RootID] = true
		}
	}
	after := ""
	for {
		page, err := s.store.CommandSessions(ctx, "", after, rebuildPage)
		if err != nil {
			return err
		}
		for _, record := range page {
			after = record.SessionID
			if record.State != storage.CommandSessionReady {
				continue
			}
			reason := s.rebuildSession(ctx, client, record, bound[record.RootID])
			if reason == "" {
				continue
			}
			if _, err = s.store.UpdateCommandSession(ctx, record.SessionID, record.Revision, record.Generation, storage.CommandSessionUnrebuildable, reason, s.now()); err != nil && !errors.Is(err, storage.ErrConflict) {
				return err
			}
		}
		if len(page) < rebuildPage {
			return nil
		}
	}
}

func (s *Service) rebuildSession(ctx context.Context, client *worker.Client, record storage.CommandSession, rootBound bool) string {
	if !rootBound {
		return "ROOT_UNAVAILABLE"
	}
	launch := new(pb.CommandLaunchSpec)
	if proto.Unmarshal(record.Launch, launch) != nil {
		return "LAUNCH_UNREADABLE"
	}
	session, err := client.CreateCommandSession(ctx, &pb.CreateCommandSessionRequest{SessionId: record.SessionID, RootId: record.RootID, WorkspaceId: record.WorkspaceID, Kind: pb.CommandSessionKind_COMMAND_SESSION_KIND_NON_INTERACTIVE_COMMAND, Launch: launch})
	if err != nil || session == nil {
		return "REBUILD_FAILED"
	}
	if session.Generation != record.Generation {
		return "GENERATION_CHANGED"
	}
	if [32]byte(sha256.Sum256(record.Launch)) != record.LaunchSHA256 || string(session.LaunchSha256) != string(record.LaunchSHA256[:]) {
		return "LAUNCH_DIGEST_CHANGED"
	}
	return ""
}

// payloads resolves immutable Host-owned content. It independently re-checks
// the digest; it never accepts a browser token or a caller-supplied body.
type payloads struct{ store *storage.Store }

func (p payloads) Resolve(ctx context.Context, workspace, reference string) ([]byte, error) {
	record, err := p.store.AutomationPayload(ctx, workspace, reference)
	if err != nil {
		return nil, err
	}
	digest := sha256.Sum256(record.Payload)
	if digest != record.SHA256 || hex.EncodeToString(digest[:]) != reference {
		return nil, storage.ErrCorrupt
	}
	return record.Payload, nil
}

// Verify re-checks a stored authorization at dispatch time without any browser
// credential: the device must still exist, be unrevoked, keep the epoch the
// grant recorded, and still hold automation:manage for this exact workspace
// and execution host.
func (s *Service) Verify(ctx context.Context, authorization automation.Authorization, config *pb.AutomationPlanConfig) error {
	if config == nil || config.Target == nil {
		return automation.ErrAuthorization
	}
	grant, epoch, revoked, err := s.store.AutomationGrant(ctx, authorization.AuthorizationID)
	if err != nil {
		return automation.ErrAuthorization
	}
	if grant.PrincipalID != authorization.PrincipalID || revoked != 0 || epoch != grant.DeviceEpoch {
		return automation.ErrAuthorization
	}
	scopes, err := auth.DecodeScopes(grant.Scopes)
	if err != nil {
		return automation.ErrAuthorization
	}
	if !auth.Permits(scopes, []auth.Scope{{Permission: ScopeManage, WorkspaceID: config.WorkspaceId, ExecutionHostID: config.Target.ExecutionHostId}}) {
		return automation.ErrAuthorization
	}
	return nil
}

// Supports answers from the Host's own definition first: an unrebuildable or
// unknown target is explicitly unsupported, not an indefinite wait.
func (s *Service) Supports(ctx context.Context, target *pb.AutomationTarget) (automation.TargetStatus, error) {
	if target == nil || target.ExecutionHostId != s.options.HostID {
		return automation.TargetStatus{State: automation.TargetUnsupported}, nil
	}
	// An agent target names a canvas node, not a session this Host froze, so
	// there is no stored definition to answer from: the Runtime that owns the
	// terminal is the only thing that can say whether it is writable.
	if automation.AgentTarget(target) {
		_, inner, _ := s.current()
		if inner == nil {
			return automation.TargetStatus{State: automation.TargetUnknown}, ErrUnsupported
		}
		return inner.Supports(ctx, target)
	}
	record, err := s.store.CommandSession(ctx, target.SessionId)
	if errors.Is(err, storage.ErrNotFound) {
		return automation.TargetStatus{State: automation.TargetUnsupported}, nil
	}
	if err != nil {
		return automation.TargetStatus{State: automation.TargetUnknown}, err
	}
	if record.State != storage.CommandSessionReady || record.ExecutionHostID != s.options.HostID || record.Generation != target.Generation {
		return automation.TargetStatus{State: automation.TargetUnsupported, Generation: record.Generation}, nil
	}
	_, inner, _ := s.current()
	if inner == nil {
		return automation.TargetStatus{State: automation.TargetUnknown}, ErrUnsupported
	}
	return inner.Supports(ctx, target)
}

func (s *Service) Dispatch(ctx context.Context, run *pb.AutomationRun) (*pb.AutomationReceipt, error) {
	_, inner, _ := s.current()
	if inner == nil {
		return nil, ErrUnsupported
	}
	return inner.Dispatch(ctx, run)
}

func (s *Service) Lookup(ctx context.Context, run *pb.AutomationRun) (*pb.AutomationReceipt, error) {
	_, inner, _ := s.current()
	if inner == nil {
		return nil, ErrUnsupported
	}
	return inner.Lookup(ctx, run)
}

var (
	_ automation.Dispatcher = (*Service)(nil)
	_ automation.Authorizer = (*Service)(nil)
)

File: apps/desktop/src/core/schedule/service.ts
import { createHash } from "node:crypto";
import {
  AutomationCommandSessionState,
  type AutomationPlanConfig,
  type AutomationTarget,
  type CommandLaunchSpec,
} from "./types";

import { decodeScopes, encodeScopes, permits, scope } from "../identity/scopes";
import type { Scope } from "../identity/scopes";
import type { IdentityStore } from "../identity/store";
import {
  ScheduleError,
  agentTarget,
  big,
  configurationHash,
  invalid,
  num,
} from "./plan";
import {
  launchSpecToJson,
  planConfigFromJson,
  planConfigToJson,
  storedJson,
} from "./json";
import {
  type Authorization,
  type Authorizer,
  type PlanSnapshot,
  type RunSnapshot,
  ScheduleEngine,
} from "./engine";
import {
  COMMAND_SESSION_READY,
  type CommandSessionRecord,
  ScheduleStore,
  conflict,
  notFound,
} from "./store";

/**
 * 已认证调用方与调度内核之间的那一层。
 *
 * 移植自 `apps/host/internal/automationhost/{api,service}.go`。它管四件内核不该
 * 管的事：**这次调用有没有权限**、**载荷由谁存**、**授权当时的授权位记在哪**，
 * 以及**投递时拿什么重新核对**。
 *
 * 两条从 Host 原样带过来的规矩：
 *
 *   * **载荷的引用与摘要由这一层算，客户端写什么都被覆盖**。否则一个计划可以
 *     指向它从没提交过的内容。
 *   * **投递时核的是记下来的那份授权，不是一个活会话**。计划在页面关掉之后还要
 *     跑；到时候要问的是「授权它的那台设备今天还在、还没被吊销、授权位还够吗」。
 */

export const SCOPE_READ = "automation:read";
export const SCOPE_MANAGE = "automation:manage";

/** 载荷上限，和 Host 的 `MaxAutomationPayloadBytes` 同一个数。 */
export const MAX_PAYLOAD_BYTES = 64 * 1024;

const ID_PATTERN = /^[A-Za-z0-9_.-]{1,128}$/;

/** 已验证会话自己的身份与授权位。没有一个字段来自客户端提交的内容。 */
export interface Caller {
  readonly principalId: string;
  readonly deviceId: string;
  readonly deviceEpoch: number;
  readonly workspaceId: string;
  readonly scopes: readonly Scope[];
}

export interface ScheduleServiceOptions {
  readonly store: ScheduleStore;
  readonly engine: ScheduleEngine;
  readonly identity: IdentityStore;
  readonly hostId: string;
  readonly clock?: () => number;
  /**
   * 一个终端会话现在的代数。
   *
   * 冻结一个命令会话就是把「现在跑着的这一个」记下来；代数因此必须来自终端域，
   * 而不是客户端说了算——客户端说的那个数可以是它上次看到的那个。
   */
  readonly generationOf?: (sessionId: string) => number | undefined;
}

const denied = (why = "这台设备没有这项授权"): ScheduleError =>
  new ScheduleError("authorization", why);

export class ScheduleService implements Authorizer {
  private readonly store: ScheduleStore;
  private readonly identity: IdentityStore;
  readonly engine: ScheduleEngine;
  readonly hostId: string;
  private readonly clock: () => number;
  private readonly generationOf: (sessionId: string) => number | undefined;

  constructor(options: ScheduleServiceOptions) {
    this.store = options.store;
    this.engine = options.engine;
    this.identity = options.identity;
    this.hostId = options.hostId;
    this.clock = options.clock ?? (() => Date.now());
    this.generationOf = options.generationOf ?? (() => undefined);
  }

  /* -------------------------------- 权限 ----------------------------------- */

  private authorize(caller: Caller, permission: string): void {
    if (
      caller.principalId === "" ||
      caller.deviceId === "" ||
      caller.deviceEpoch <= 0 ||
      !ID_PATTERN.test(caller.workspaceId) ||
      caller.scopes.length === 0
    ) {
      throw denied("调用方身份不完整");
    }
    if (
      !permits(caller.scopes, [
        scope(permission, caller.workspaceId, this.hostId),
      ])
    ) {
      throw denied();
    }
  }

  private authorization(caller: Caller): Authorization {
    return {
      principalId: caller.principalId,
      authorizationId: caller.deviceId,
    };
  }

  /** 记下这台设备授权时握着的授权位。日后的投递核的是这一行。 */
  private recordGrant(caller: Caller): void {
    const now = this.clock();
    this.store.putGrant({
      authorizationId: caller.deviceId,
      principalId: caller.principalId,
      deviceId: caller.deviceId,
      deviceEpoch: caller.deviceEpoch,
      scopes: encodeScopes(caller.scopes),
      atMs: now,
    });
  }

  /**
   * 投递时重新核一次，不碰任何浏览器凭据。
   *
   * 四件事都要成立：设备还在、没被吊销、纪元还是授权时那一个、而且它仍然对这个
   * 工作空间与这台执行主机握着 `automation:manage`。
   */
  async verify(
    authorization: Authorization,
    config: AutomationPlanConfig,
  ): Promise<void> {
    if (config.target === undefined) throw denied("计划没有目标");
    const grant = this.store.grant(authorization.authorizationId);
    if (
      grant === undefined ||
      grant.principalId !== authorization.principalId
    ) {
      throw denied("找不到当初那次授权");
    }
    const device = this.identity.transaction((tx) => tx.device(grant.deviceId));
    if (
      device === undefined ||
      device.revokedAtMs !== 0 ||
      device.epoch !== grant.deviceEpoch
    ) {
      throw denied("授权它的那台设备已经失效");
    }
    let scopes: Scope[];
    try {
      scopes = decodeScopes(grant.scopes);
    } catch {
      throw denied("授权记录读不出来");
    }
    if (
      !permits(scopes, [
        scope(SCOPE_MANAGE, config.workspaceId, config.target.executionHostId),
      ])
    ) {
      throw denied();
    }
  }

  /* ------------------------------- 命令会话 -------------------------------- */

  /**
   * 冻结一个非交互命令会话。
   *
   * 同进程之后这里不再有 Rust Worker：一个命令会话就是**一个纯终端节点的会话**
   * 加上它被冻结时的代数。写入只进那个 shell pane，永远不进一个 Agent 的 pane。
   */
  defineCommandSession(
    caller: Caller,
    sessionId: string,
    rootPath: string,
    launch: CommandLaunchSpec,
  ): CommandSessionRecord {
    this.authorize(caller, SCOPE_MANAGE);
    if (
      !ID_PATTERN.test(sessionId) ||
      rootPath === "" ||
      rootPath.includes("\u0000") ||
      launch.executable === ""
    ) {
      throw invalid("命令会话的定义不完整");
    }
    // 代数来自终端域。没在跑的会话冻结不了：冻结的是「这一个进程」，而现在
    // 没有哪一个进程可以被指着说就是它。
    const generation = this.generationOf(sessionId);
    if (generation === undefined || generation <= 0) {
      throw new ScheduleError("unsupported", "这个终端会话现在没有在跑");
    }
    // 走一遍编码：一份编不出来的启动定义在这里被拒，而不是在第一次该跑的时候。
    // 冻结的是**规范 JSON**（0020），所以身份摘要也按它算。
    const frozen = storedJson(launchSpecToJson(launch));
    const rootId = `root-${createHash("sha256")
      .update(`${caller.workspaceId}\u0000${rootPath}`)
      .digest("hex")
      .slice(0, 32)}`;
    const now = Math.max(1, this.clock());
    return this.store.transact(() => {
      this.store.putCommandRoot(rootId, caller.workspaceId, rootPath, now);
      const record = this.store.putCommandSession({
        sessionId,
        rootId,
        workspaceId: caller.workspaceId,
        executionHostId: this.hostId,
        launch,
        launchSha256: createHash("sha256").update(frozen, "utf8").digest(),
        generation,
        state: COMMAND_SESSION_READY,
        reasonCode: "",
        revision: 1,
        createdAtMs: now,
        updatedAtMs: now,
      });
      this.recordGrant(caller);
      return record;
    });
  }

  listCommandSessions(
    caller: Caller,
    after: string,
    limit: number,
  ): {
    sessions: CommandSessionRecord[];
    rootPaths: Map<string, string>;
    nextId: string;
    hasMore: boolean;
  } {
    this.authorize(caller, SCOPE_READ);
    const page = this.store.listCommandSessions(
      caller.workspaceId,
      after,
      limit <= 0 || limit > 200 ? 50 : limit,
    );
    const rootPaths = new Map<string, string>();
    for (const session of page.sessions) {
      const path = this.store.commandRoot(session.rootId);
      if (path !== undefined) rootPaths.set(session.rootId, path);
    }
    return { ...page, rootPaths };
  }

  /* --------------------------------- 计划 ---------------------------------- */

  /**
   * 存一份计划配置与它不可变的载荷。
   *
   * 载荷的引用和摘要由这一层算，客户端写在那两个字段里的东西一律被覆盖——否则
   * 一个计划可以指向它从没提交过的内容。
   */
  async define(
    caller: Caller,
    planId: string,
    config: AutomationPlanConfig,
    payload: Uint8Array,
    expectedRevision: number,
  ): Promise<PlanSnapshot> {
    this.authorize(caller, SCOPE_MANAGE);
    if (
      !ID_PATTERN.test(planId) ||
      config.target === undefined ||
      payload.byteLength > MAX_PAYLOAD_BYTES
    ) {
      throw invalid("计划定义不完整");
    }
    // 深拷贝一份再改：调用方交来的那个对象不属于这里。走 JSON 一圈同时也验了
    // 「这份配置编得出来」——编不出来的配置在这里被拒，而不是在第一次写库时。
    const prepared = planConfigFromJson(planConfigToJson(config));
    prepared.workspaceId = caller.workspaceId;
    const target = prepared.target as AutomationTarget;
    if (target.executionHostId !== this.hostId) {
      throw new ScheduleError("unsupported", "这个计划指向另一台执行主机");
    }
    // 命令目标指名一个这个 core 冻结过的会话，所以定义在这里就核。Agent 目标
    // 指名的是一个画布节点，那个会话合法地来来去去，没有什么可以提前查——投递方
    // 在每次探测和每次写入时重新核节点的身份。
    if (!agentTarget(target)) {
      const record = this.store.commandSession(target.sessionId);
      if (record === undefined) throw notFound("找不到这个命令会话");
      if (
        record.workspaceId !== caller.workspaceId ||
        record.state !== COMMAND_SESSION_READY
      ) {
        throw new ScheduleError("unsupported", "这个命令会话现在用不了");
      }
      if (num(target.generation) === 0) {
        target.generation = big(record.generation);
      }
      if (num(target.generation) !== record.generation) throw conflict();
    }
    const digest = createHash("sha256").update(payload).digest();
    prepared.payloadRef = digest.toString("hex");
    prepared.payloadSha256 = digest;
    this.store.putPayload(
      caller.workspaceId,
      prepared.payloadRef,
      payload,
      digest,
      Math.max(1, this.clock()),
    );
    this.recordGrant(caller);
    return this.engine.define(
      this.authorization(caller),
      planId,
      prepared,
      expectedRevision,
    );
  }

  /**
   * 读回一个计划被定义时用的那份 stdin / 提示词。
   *
   * 改一个计划要重发整份配置，所以表单必须从它真正带着的那份载荷开始：让人重打
   * 一遍是负担，而悄悄换成空的等于把计划改成了什么都不输入。
   */
  payload(caller: Caller, planId: string): Uint8Array {
    this.authorize(caller, SCOPE_MANAGE);
    if (!ID_PATTERN.test(planId)) throw invalid("计划标识不合法");
    const snapshot = this.engine.getPlan(caller.workspaceId, planId);
    const config = snapshot.plan.config;
    if (config === undefined) throw invalid();
    const record = this.store.payload(caller.workspaceId, config.payloadRef);
    // 计划报了一个摘要；和它对不上的字节不是这个计划的载荷，交出去就等于让一次
    // 编辑投递出读的人从没看过的东西。
    const digest = createHash("sha256").update(record.payload).digest();
    if (
      digest.toString("hex") !==
      Buffer.from(config.payloadSha256).toString("hex")
    ) {
      throw invalid("载荷与计划记录的摘要对不上");
    }
    return record.payload;
  }

  async activate(
    caller: Caller,
    planId: string,
    expectedRevision: number,
    configVersion: number,
    digest: Uint8Array,
  ): Promise<PlanSnapshot> {
    this.authorize(caller, SCOPE_MANAGE);
    this.recordGrant(caller);
    return this.engine.activate(
      this.authorization(caller),
      caller.workspaceId,
      planId,
      expectedRevision,
      configVersion,
      digest,
    );
  }

  async pause(
    caller: Caller,
    planId: string,
    expectedRevision: number,
  ): Promise<PlanSnapshot> {
    this.authorize(caller, SCOPE_MANAGE);
    return this.engine.pause(
      this.authorization(caller),
      caller.workspaceId,
      planId,
      expectedRevision,
    );
  }

  async runNow(
    caller: Caller,
    planId: string,
    expectedRevision: number,
  ): Promise<RunSnapshot> {
    this.authorize(caller, SCOPE_MANAGE);
    this.recordGrant(caller);
    return this.engine.runNow(
      this.authorization(caller),
      caller.workspaceId,
      planId,
      expectedRevision,
    );
  }

  listPlans(
    caller: Caller,
    after: string,
    limit: number,
  ): { plans: PlanSnapshot[]; nextId: string; hasMore: boolean } {
    this.authorize(caller, SCOPE_READ);
    return this.engine.listPlans(caller.workspaceId, after, limit);
  }

  listRuns(
    caller: Caller,
    planId: string,
    after: string,
    limit: number,
  ): { runs: RunSnapshot[]; nextId: string; hasMore: boolean } {
    this.authorize(caller, SCOPE_READ);
    return this.engine.listRuns(caller.workspaceId, planId, after, limit);
  }

  /** 计划快照要带上配置摘要：没有它的快照激活不了，所以它不是一个完整的快照。 */
  configDigest(snapshot: PlanSnapshot): Uint8Array {
    const config = snapshot.plan.config;
    if (config === undefined) throw invalid();
    return configurationHash(config);
  }
}

export { AutomationCommandSessionState };
